Add documentation for tracing SDK (#15541)
mlflow/mlflow: docs/docs/tracing/minimal-sdk.mdx

---
description: MLflow Tracing SDK (mlflow-tracing) is a lightweight package that includes only the essential functionality for tracing and monitoring your GenAI applications.
sidebar_position: 5
sidebar_label: Lightweight SDK
---

import { Card, CardGroup, TitleCard } from "@site/src/components/Card";
import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

# Lightweight Tracing SDK Optimized for Production Usage

> MLflow offers a lightweight tracing SDK package called `mlflow-tracing` that includes only the essential functionality for tracing and monitoring your GenAI applications. This package is designed for production environments where minimizing dependencies and deployment size is crucial.

Reviewer: ```suggestion
MLflow offers a lightweight tracing SDK package called `mlflow-tracing` that includes only the essential functionality for tracing and monitoring of your GenAI applications. This package is designed for production environments where minimizing dependencies and deployment size is ideal.
```
Something about the word "crucial" triggers something in my hind-brain. "OMG, so, using the full MLflow package is WRONG?" Smaller package is 'ideal' I think

Author: haha totally makes sense!

## Why Use the Lightweight SDK?

<CardGroup cols={2}>
    <TitleCard
        title="🏎️ Faster Deployment"
        description="The package size and dependencies are significantly smaller than the full MLflow package, allowing for faster deployment times in dynamic environments such as Docker containers, serverless functions, and cloud-based applications."
    />
    <TitleCard
        title="🔧 Simplified Dependency Management"

>         description="A smaller set of dependencies means less work keeping up with dependency updates, security patches, and breaking changes from upstream libraries."

Reviewer: ```suggestion
        description="A smaller set of dependencies means less work keeping up with dependency updates, security patches and potential breaking changes from upstream libraries. It also reduces the chances of dependency collisions and incompatibilities."
```

    />
    <TitleCard
        title="📦 Portability"

>         description="With the less number of dependencies, MLflow Tracing can be easily deployed across different environments and platforms, without worrying about compatibility issues."

Reviewer: ```suggestion
        description="With a fewer number of dependencies, MLflow Tracing can be seamlessly deployed across differing environments and platforms, without worrying about compatibility issues."
```

    />
    <TitleCard
        title="🔒 Fewer Security Risks"
        description="Each dependency potentially introduces security vulnerabilities. By reducing the number of dependencies, MLflow Tracing minimizes the attack surface and reduces the risk of security breaches."
    />
</CardGroup>

## Installation

Install the lightweight SDK using pip:

```bash
pip install mlflow-tracing
```

:::warning

> Do not install the full `mlflow` package together with the lightweight SDK, as this may cause version conflicts.

Reviewer: ```suggestion
Do not install the full `mlflow` package together with the lightweight `mlflow-tracing` SDK, as this may cause version conflicts and namespace resolution issues.
```

on a remote MLflow server. If you haven't set up a remote MLflow server yet, please refer to the [Choose Your Backend](#choose-your-backend) section below.

```python
import mlflow
import openai

# Set the tracking URI (e.g., Databricks, SageMaker, or self-hosted server)
mlflow.set_tracking_uri("<your-tracking-uri>")
mlflow.set_experiment("<your-experiment-name>")

# Enable auto-tracing for OpenAI
mlflow.openai.autolog()

# Use OpenAI as usual
client = openai.OpenAI()
response = client.chat.completions.create(
    model="gpt-4", messages=[{"role": "user", "content": "What is MLflow?"}]
)
```

## Choose Your Backend

The lightweight SDK works with various observability platforms. Choose your favorite one and follow the instructions to set up your tracing backend.

<Tabs>
  <TabItem value="databricks-monitoring" label="Databricks">
    <div class="flex-column">
      <div class="flex-row">
        <div class="flex-item">

          [Databricks Lakehouse Monitoring for GenAI](https://docs.databricks.com/aws/en/generative-ai/agent-evaluation/monitoring) is a go-to solution for monitoring your

>           GenAI application with MLflow Tracing. It provides an instant access to a fully functional monitoring dashboard for operational excellence and quality analysis.

Reviewer: ```suggestion
          GenAI application with MLflow Tracing. It provides access to a robust, fully functional monitoring dashboard for operational excellence and quality analysis.
```

          Lakehouse Monitoring for GenAI can be used regardless of whether your application is hosted on Databricks or not.

          [Sign up for free](https://docs.databricks.com/aws/en/getting-started/free-trial) and [get started in a minute](https://docs.databricks.com/aws/en/generative-ai/agent-evaluation/monitoring) to run your GenAI application with complete observability.
        </div>

        <div class="flex-item padding-md">
          ![Monitoring](https://assets.docs.databricks.com/_static/images/generative-ai/monitoring/monitoring-hero.gif)
        </div>
      </div>
    </div>

  </TabItem>
  <TabItem value="self-hosting" label="Self-hosting">
    <div class="flex-column">
      <div class="flex-row">
        <div class="flex-item">

>           MLflow is a **fully open-source project**, allowing you to self-host your own MLflow server and keep your data private. This is a great option if you want to have full control over your data and infrastructure.

Reviewer: ```suggestion
          MLflow is a **fully open-source project**, allowing you to self-host your own MLflow server in your own infrastructure. This is a great option if you want to have full control over your data and are restricted in using cloud-based services.
```
Let's not mention "keeping data private". This implies that storage in cloud based services of MLflow isn't secure, which is not true. Most large-scale self-hosting is due to restrictions about keeping service on on-prem data centers rather than a fear that a provider can't keep data isolated.

          In self-hosting mode, you will be responsible for running the tracking server instance

>           and scaling it to your needs. We **strongly recommend** using SQL-based tracking server

Reviewer: ```suggestion
          and scaling it to your needs. We **strongly recommend** using a SQL-based tracking server
```

          on top of the performant database of your choice to minimize the operational overhead
          and high availability. Please refer to the [tracking server setup guide](/tracking#tracking-setup) for more guidance.

        </div>
        <div class="flex-item padding-md">
          ![OSS Tracing](/images/llms/tracing/tracing-top.gif)
        </div>
      </div>
    </div>
  </TabItem>
  <TabItem value="other" label="OpenTelemetry">
    <div class="flex-column">
      <div class="flex-row">
        <div class="flex-item">

>           MLflow Tracing is **built on OpenTelemetry tracing spec**, an industry-standard framework for observability, making it vendor neutral solution for monitoring your GenAI applications.

Reviewer: ```suggestion
          MLflow Tracing is **built on the OpenTelemetry tracing spec**, an industry-standard framework for observability, making it a vendor-neutral solution for monitoring your GenAI applications.
```

        </div>
      </div>
    </div>
  </TabItem>
  <TabItem value="nebius" label="Nebius">
    <div class="flex-column">
      <div class="flex-row">
        <div class="flex-item">

          [Nebuis](https://nebius.com/services/managed-mlflow), a cutting-edge cloud platform for GenAI explorers, offers a fully managed MLflow server. Combining the powerful GPU infrastructure of Nebuis for training and hosting LLMs / foundation models with the observability capabilities of MLflow Tracing, Nebius serves as a comprehensive platform for AI/ML developers.

          Refer to the [Nebius documentation](https://nebius.com/services/managed-mlflow) for more details about the managed MLflow service.

        </div>
        <div class="flex-item padding-md">
          <img src="/images/logos/nebius-logo.png" alt="Nebuis Logo" width="80%" />
        </div>
      </div>
    </div>
  </TabItem>
</Tabs>

## Supported Features

The lightweight SDK includes all essential tracing functionalities for monitoring your GenAI applications. Click the cards below to learn more about each supported feature.

<CardGroup cols={1}>
    <TitleCard
        title="⚡️ Automatic Tracing for 15+ AI libraries"

>         description="MLflow Tracing SDK supports one-line integration with 15+ AI libraries including OpenAI, LangChain, LlamaIndex, Anthropic, Amazon Bedrock, DSPy, and more. This

Reviewer: Instead of 'number dropping' here, can we just say that we support all of the most popular as well as anything that conforms to the OpenAI spec?

        automatic tracing capability allows you to monitor your GenAI application with MLflow Tracing with minimal effort and switch between different libraries easily."
        link="/tracing/integrations/"
    />
    <TitleCard
        title="⚙️ Manual Instrumentation"
        description="MLflow Tracing SDK provides a simple and intuitive API for manually
        instrumenting your GenAI application. Manual instrumentation and automatic tracing can

>         be used together, allowing you to trace an advanced applications with custom code and
>         fine-grained control over the tracing behavior."

Reviewer: ```suggestion
        be used together, allowing you to trace advanced applications containing custom code and have
        fine-grained control over the tracing behavior."
```

        link="/tracing/api/manual-instrumentation"
    />
    <TitleCard
        title="✍ Tagging and Filtering Traces"

>         description="By annotating traces with custom tags, you can add more context to your traces for grouping and searching for them later. This is useful when you want to trace

Reviewer: ```suggestion
        description="By annotating traces with custom tags, you can add more context to your traces to group them and simplify the process of searching for them later. This is useful when you want to trace
```
